File: apps/api/app/api/routes/feedback.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.schemas.feedback import (
    FeedbackCreate,
    FeedbackResponse,
    FeedbackUpdate,
)
from app.services.feedback import FeedbackService


router = APIRouter(
    prefix="/feedback",
)


@router.get(
    "",
    response_model=list[FeedbackResponse],
)
def get_all_feedback(
    submission_id: UUID | None = Query(
        default=None, description="Filter by submission ID"
    ),
    reviewer_id: UUID | None = Query(
        default=None, description="Filter by reviewer ID"
    ),
    db: Session = Depends(get_db),
):
    service = FeedbackService(db)
    return service.get_all(
        submission_id=submission_id,
        reviewer_id=reviewer_id,
    )


@router.get(
    "/{feedback_id}",
    response_model=FeedbackResponse,
)
def get_feedback(
    feedback_id: UUID,
    db: Session = Depends(get_db),
):
    service = FeedbackService(db)
    feedback = service.get_by_id(feedback_id)
    if not feedback:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Feedback not found",
        )
    return feedback


@router.post(
    "",
    response_model=FeedbackResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_feedback(
    data: FeedbackCreate,
    db: Session = Depends(get_db),
):
    service = FeedbackService(db)
    return service.create(data)


@router.patch(
    "/{feedback_id}",
    response_model=FeedbackResponse,
)
def update_feedback(
    feedback_id: UUID,
    data: FeedbackUpdate,
    db: Session = Depends(get_db),
):
    service = FeedbackService(db)
    return service.update(feedback_id, data)


@router.delete(
    "/{feedback_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_feedback(
    feedback_id: UUID,
    db: Session = Depends(get_db),
):
    service = FeedbackService(db)
    service.delete(feedback_id)

File: apps/web/lib/api/feedback.ts
import { apiClient } from "@/lib/api/client";
import { Feedback, FeedbackCreate, FeedbackUpdate } from "@/types/fellowship";

export async function fetchFeedback(params?: {
  submission_id?: string;
  reviewer_id?: string;
}): Promise<Feedback[]> {
  const query = new URLSearchParams();
  if (params?.submission_id) query.append("submission_id", params.submission_id);
  if (params?.reviewer_id) query.append("reviewer_id", params.reviewer_id);
  const qs = query.toString() ? `?${query.toString()}` : "";
  return apiClient<Feedback[]>(`/feedback${qs}`);
}

export async function fetchFeedbackById(id: string): Promise<Feedback> {
  return apiClient<Feedback>(`/feedback/${id}`);
}

export async function createFeedback(data: FeedbackCreate): Promise<Feedback> {
  return apiClient<Feedback>("/feedback", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export async function updateFeedback(
  id: string,
  data: FeedbackUpdate
): Promise<Feedback> {
  return apiClient<Feedback>(`/feedback/${id}`, {
    method: "PATCH",
    body: JSON.stringify(data),
  });
}

export async function deleteFeedback(id: string): Promise<void> {
  return apiClient<void>(`/feedback/${id}`, {
    method: "DELETE",
  });
}
